metakb/harvesters/civic.py

```diff
@@ -0,0 +1,377 @@
+from .base import Harvester
+from metakb import PROJECT_ROOT
+from civicpy import civic as civicpy
+import json
+
+
+class CIViC(Harvester):
+    """A class for the CIViC harvester."""
+
+    def harvest(self):
+        """Retrieve and store records from CIViC."""
+        try:
+            # civicpy.load_cache(on_stale='ignore')
+            evidence = self._harvest_evidence()
+            genes = self._harvest_genes()
+            variants = self._harvest_variants()
+            assertions = self._harvest_assertions()
+            self._create_json(evidence, genes, variants, assertions)
+            print('CIViC Harvester was successful.')
+            return True
+        except:  # noqa: E722 # TODO: Add specific exception error
+            print('CIViC Harvester was not successful.')
+            return False
+
+    def _create_json(self, evidence, genes, variants, assertions):
+        """Create a composite json file containing evidence, genes, variants,
+        and assertions.
+        """
+        composite_dict = {
+            'evidence': evidence,
+            'genes': genes,
+            'variants': variants,
+            'assertions': assertions
+        }
+
+        # Create composite json
+        with open(f'{PROJECT_ROOT}/data/civic/civic_harvester.json',
+                  'w+') as f:
+            json.dump(composite_dict, f)
+            f.close()
+
+        # Create individual json for evidence, genes, variants, and assertions
+        data = ['evidence', 'genes', 'variants', 'assertions']
+        for d in data:
+            with open(f'{PROJECT_ROOT}/data/civic/{d}.json', 'w+') as f:
+                f.write(json.dumps(composite_dict[d]))
+                f.close()
+
+    def _harvest_evidence(self):
+        """Harvest evidence."""
+        evidence_classes = civicpy.get_all_evidence()
+        evidence = list()
+
+        for ev in evidence_classes:
+            ev_record = self._evidence_item(ev, is_evidence=True)
+            evidence.append(ev_record)
+        return evidence
+
+    def _harvest_genes(self):
+        """Harvest genes."""
+        genes = civicpy.get_all_genes()
+        genes_list = list()
+        for gene in genes:
+            g = self._harvest_gene(gene)
+            genes_list.append(g)
+        return genes_list
+
+    def _harvest_variants(self):
+        """Harvest variants."""
+        variants = civicpy.get_all_variants()
+        variants_list = list()
+
+        for variant in variants:
+            v = self._harvest_variant(variant)
+            variants_list.append(v)
+        return variants_list
+
+    def _harvest_assertions(self):
+        """Harvest assertions."""
+        assertions = civicpy.get_all_assertions()
+        assertions_list = list()
+
+        for assertion in assertions:
+            a = self._harvest_assertion(assertion)
+            assertions_list.append(a)
+        return assertions_list
+
+    def _harvest_gene(self, gene):
+        g = {
```

Comment:
We could use a doc string on these `_harvest_<singular>()` methods.   The code looks fine, but I'm unclear as to the intent.

As I read the code on first pass, it seems they are defining a normalized structure for the entity.
In some cases, like this _harvest_gene, a new object is defined `g`.
In other cases, like _harvest_variant, the class variable is decorated with additional fields `v_extra`

Reply:
A lot of the harvester code was redundant at first. All methods without 'harvest' in them (i.e. `_evidence_item`, `_variant`, `_assertion`) contain this redundant code that is used in all four CIViC records. Variants and Assertions contain extra attributes  (`v_extra` and `a_extra`) not found in the other records.